File: Sources/AstronomyEngineMobile/AstronomyEngine.swift
import CAstronomyEngine
import Foundation

public enum AstronomyBody: Int32, CaseIterable, Sendable {
    case mercury = 0
    case venus
    case mars
    case jupiter
    case saturn
    case uranus
    case neptune
    case pluto
    case sun
    case moon
}

public enum LunarNodeConvention: Int32, Sendable {
    case mean = 0
    case trueOsculating = 1
}

public struct EclipticLongitude: Equatable, Sendable {
    public let degrees: Double
    public let maximumErrorDegrees: Double

    public init(degrees: Double, maximumErrorDegrees: Double) {
        self.degrees = degrees
        self.maximumErrorDegrees = maximumErrorDegrees
    }
}

public struct AstronomyEngineMetadata: Equatable, Sendable {
    public let wrapperVersion: String
    public let upstreamVersion: String
    public let coordinateContract: String
    public let trueNodeContract: String
}

public enum AstronomyEngineError: Int32, Error, Equatable, Sendable {
    case invalidBody = 1
    case invalidTime = 2
    case calculationError = 3
    case invalidArgument = 4
    case unknown = -1
}

public enum AstronomyEngine {
    public static var wrapperVersion: String { String(cString: ame_wrapper_version()) }
    public static var upstreamVersion: String { String(cString: ame_upstream_version()) }
    public static var metadata: AstronomyEngineMetadata {
        AstronomyEngineMetadata(
            wrapperVersion: wrapperVersion,
            upstreamVersion: upstreamVersion,
            coordinateContract: String(cString: ame_coordinate_contract()),
            trueNodeContract: String(cString: ame_true_node_contract())
        )
    }

    public static func longitude(
        of body: AstronomyBody,
        at date: Date
    ) throws -> EclipticLongitude {
        try map(ame_ecliptic_longitude(
            ame_body_t(rawValue: UInt32(body.rawValue)),
            epochMilliseconds(date)
        ))
    }

    public static func lunarAscendingNode(
        convention: LunarNodeConvention,
        at date: Date
    ) throws -> EclipticLongitude {
        try map(ame_lunar_ascending_node(
            ame_node_convention_t(rawValue: UInt32(convention.rawValue)),
            epochMilliseconds(date)
        ))
    }

    private static func epochMilliseconds(_ date: Date) -> Int64 {
        Int64((date.timeIntervalSince1970 * 1_000.0).rounded())
    }

    private static func map(_ result: ame_longitude_result_t) throws -> EclipticLongitude {
        guard result.status == AME_SUCCESS else {
            throw AstronomyEngineError(rawValue: Int32(result.status.rawValue)) ?? .unknown
        }
        return EclipticLongitude(
            degrees: result.degrees,
            maximumErrorDegrees: result.maximum_error_degrees
        )
    }
}

File: android/src/main/kotlin/com/nyra/astronomy/mobile/AstronomyEngine.kt
package com.nyra.astronomy.mobile

enum class AstronomyBody(internal val nativeCode: Int) {
    Mercury(0),
    Venus(1),
    Mars(2),
    Jupiter(3),
    Saturn(4),
    Uranus(5),
    Neptune(6),
    Pluto(7),
    Sun(8),
    Moon(9),
}

enum class LunarNodeConvention(internal val nativeCode: Int) {
    Mean(0),
    TrueOsculating(1),
}

data class EclipticLongitude(
    val degrees: Double,
    val maximumErrorDegrees: Double,
)

data class AstronomyEngineMetadata(
    val wrapperVersion: String,
    val upstreamVersion: String,
    val coordinateContract: String,
    val trueNodeContract: String,
)

class AstronomyCalculationException(
    val statusCode: Int,
) : IllegalArgumentException("Astronomy calculation failed with status $statusCode")

object AstronomyEngine {
    val wrapperVersion: String get() = NativeBridge.wrapperVersion()
    val upstreamVersion: String get() = NativeBridge.upstreamVersion()
    val metadata: AstronomyEngineMetadata
        get() = AstronomyEngineMetadata(
            wrapperVersion = wrapperVersion,
            upstreamVersion = upstreamVersion,
            coordinateContract = NativeBridge.coordinateContract(),
            trueNodeContract = NativeBridge.trueNodeContract(),
        )

    fun longitude(
        body: AstronomyBody,
        epochMilliseconds: Long,
    ): EclipticLongitude = NativeBridge.longitude(body.nativeCode, epochMilliseconds).toResult()

    fun lunarAscendingNode(
        convention: LunarNodeConvention,
        epochMilliseconds: Long,
    ): EclipticLongitude = NativeBridge.lunarNode(
        convention.nativeCode,
        epochMilliseconds,
    ).toResult()

    private fun DoubleArray.toResult(): EclipticLongitude {
        require(size == 3) { "Native result must contain status, degrees, and error" }
        val status = this[0].toInt()
        if (status != 0) throw AstronomyCalculationException(status)
        return EclipticLongitude(degrees = this[1], maximumErrorDegrees = this[2])
    }
}

internal object NativeBridge {
    init {
        System.loadLibrary("astronomy_engine_mobile")
    }

    external fun wrapperVersion(): String
    external fun upstreamVersion(): String
    external fun coordinateContract(): String
    external fun trueNodeContract(): String
    external fun longitude(bodyCode: Int, epochMilliseconds: Long): DoubleArray
    external fun lunarNode(conventionCode: Int, epochMilliseconds: Long): DoubleArray
}
